File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.generatePermu([], ans, nums)
        return ans
    def generatePermu(self, curr, ans, nums):
        if len(curr) == len(nums):
            ans.append(curr[:])
            return
        for x in nums:
            if x not in curr:
                curr.append(x)
                self.generatePermu(curr, ans, nums)
                curr.pop()

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> curr = {};
        generatePermu(curr, ans, nums);
        return ans;
    }
    void generatePermu(vector<int> &curr, vector<vector<int>> &ans, vector<int> &nums){
        if(curr.size() == nums.size()){
            ans.push_back(curr);
            return;
        }
        for(int x : nums){
            if(find(curr.begin(), curr.end(), x) == curr.end()){
                curr.push_back(x);
                generatePermu(curr, ans, nums);
                curr.pop_back();
            }
        }
    }
};